### ISSUE TITLE
What is the previous performance of Rhizome?

### ISSUE THREAD

**Issue by kirbysayshi** [sebpiq/rhizome#107]
I recently deployed Rhizome at two separate events:

1) Around 35 clients using a 2014 AirPort Extreme and a 2010 macbook pro as host
2) Around 350 clients using good conference WiFi (RejectJS 2015) and a 2015 MacBook Pro

Both times, when everyone was connecting, Rhizome (also serving static files) crashed and had to be manually restarted.

Both times, approximately half of the phones in the audience had trouble connecting (this could have been due to older OS's, strange web socket support, etc, not necessarily rhizome-server itself).

What have other people's experiences been like using Rhizome? Has it been more stable than what I'm seeing? Or is this what others are seeing?

And what techniques have others used to monitor performance and profile Rhizome? So far this seems incredibly difficult because instability only manifests with lots of real clients.

And to be clear, I'm not criticizing rhizome here, just curious what others' experiences are, and if there are some known issues! Thanks for rhizome, it saved me tons of time!

**Comment by sebpiq** [comment]
Hi Andrew!

Sorry you had problems... With 350 clients I am not surprised. I made a concert with ~150 clients myself and the server had errors all the time, but I think it was due to the http server (express), not managing to keep up with sending all the files (web page, images, sounds, etc ...), and not with rhizome itself. I kept the traceback of the bug I had #95 . The server was not crashing though.

And then, with 35 clients only it is very strange that it would crash. Do you have a traceback of the bug that happened to you? I think this is **not** normal, and I haven't experienced this myself. If you can send me the traceback, or the full log, I need to look into this. 

Also, for concerts, to stay on the safe side, I run rhizome using [node-forever](https://github.com/foreverjs/forever), so in case it crashes it restarts immediately. You should do this too!!! I need to write this somewhere ...

Also, scalability is something I wasn't too concerned about until recently when we've started to make bigger concerts. But it's on my route map, and probably I'll start looking at it seriously starting next week. This includes : 

- implementing load testing for web sockets and rhizome in general (need to google about this)
- modular structure (it should be possible to deploy rhizome on a distributed architecture, i.e. several instances of rhizome talking to each other through for example redis queue)
- fixing these old bugs about some people not managing to connect to websockets #78 (this is number 1 on my list, and I will fix it next week )

This said, once again the crash you had with only 35 clients shouldn't have happened (sorry about that!!!). Can you tell me what kind of things you were doing? How messages you were sending and so on?

**Comment by sebpiq** [comment]
ping @kirbysayshi :) Any chance you would send me a traceback so I can fix the bug you had?

**Comment by kirbysayshi** [comment]
Heuy @sebpiq sorry I've been away for so long. I can post some logs, but they unfortunately do not have an actual stack trace in them. Seems like eventually the server just stalled out and became unresponsive, and then I'd ctrl+c it and restart.

The connections, log, and events are in this gist: https://gist.github.com/kirbysayshi/3803e6eb2841bf074331

One thing that my project is doing that may be different than others is that I'm using a webpage that acts as the "conductor", and whenever a new client connects, that client requests the world-state, which is provided by the conductor. So each time the state updates, every single client is sent a large json blob (https://github.com/kirbysayshi/synae-server/blob/27409a849256a9d21dc30d539d683a6f9a7089d3/frontend/index.js#L36-L374).

But that didn't seem to be the trouble it was more in getting static assets to all the clients at initial load (JS, css, some small mp3s). Once every one had "connected" at least once, and had the files cached, then sending messages seemed to work fine. But that initial connection failed often.

**Comment by sebpiq** [comment]
yes ... I don't think the blob sending should be a problem. 
I also think the issue is sending assets. I've had issues with this before. Server crashing because 100 people were dowloading a sound file at the same time. But there's nothing I can do inside rhizome to fix this. A solution would be to have a dedicated delivery server for static assets. Probably even better if on a separate machine, to not block disk access.

**Comment by kirbysayshi** [comment]
any ideas on how to actually separate these things out easily? The rhizome bin file does a lot of what appears to be magic to me, merging servers and such, and settings globals with auto-resolving addresses / ports.

I can take another look tonight but splitting the message server and assets seemed non-trivial to me.

**Comment by sebpiq** [comment]
Yes the bin does a lot of stuff but mostly to report config errors, make defining setting easy, start express server for the static assets, and so on... 

Using rhizome without the bin is actually very straightforward. I should really document the API. Something like this is enough (note that servers are optional. You dont need an OSC server if you dont use OSC) : 

```javascript
var rhizome = require('rhizome-server')
manager = new rhizome.connections.ConnectionManager({ store: '/tmp/rhizome' })
wsServer = new rhizome.websockets.Server({ port: 8000 })
oscServer = new rhizome.osc.Server({ port: 9000 })

rhizome.starter(manager, [oscServer, wsServer], function(err) {
  if (err) throw err
  console.log('started!')
})
```

Then if you have a separate server for static assets (including your html), you will have to probably run it on a separate port / separate host as rhizome. So you need to configure the rhizome client to connect to the right server (the one hosting rhizome). There is not yet an API to do that, but it's easy to trick. Here's how the current client is configured : https://github.com/sebpiq/rhizome/blob/master/lib/websockets/Client.js#L31 so by default, it takes the host/port serving the web page.
So what you need to do is before calling `rhizome.start` in your frontend code, change the config : https://github.com/sebpiq/rhizome/blob/master/lib/websockets/Client.js#L291 you can do that easily with (for example) : 

```javascript
rhizome._config.port = 8000
rhizome._config.hostname = 'myRhizomeserver.com'
```

Now, I've never ran rhizome with a different host for the assets. If you do that and run into problems (or if you succeed) let me know! That's something I want to do as well ...
